### ISSUE TITLE
SSE:  Invalid state: Controller is already closed

### ISSUE THREAD

**Issue by JasonColeyNZ** [rphlmr/react-router-hono-server#26]
Hi, loving being able to use middleware with RR7, but I have an issue with SSE and Hono, not sure what is happening.

I have been using Remix and remix-live-loader (https://github.com/moishinetzer/remix-live-loader) for sometime, and have this week making the changes for RR7, using your hono-server implementation for the middleware ability.

I am getting intermittent errors using remix-live-loader and react-router-hono-server together, it seems as though connections are not being closed when navigating back and forth.

I have a repo here to take a look at, its the example given by remix-live-loader with your hono-server added.
https://github.com/JasonColeyNZ/remix-live-loader

If you open two windows and navigate back and forth between the task lists on the left and mark and change task completion, eventually (after 10-20) changes and switching back and forth between the two tabs you will get the controller error.

Any chance you can take a look and maybe help find out what is happening?

**Comment by rphlmr** [comment]
Hello 👋

I will look at it but I'm pretty sure that you need the same implementation as https://github.com/rphlmr/drizzle-lab/blob/b07f58e8a40aaffcadb62e243810a3a179aafbad/apps/cli/visualizer/routes/watch.ts#L17C1-L31C63

```ts
export async function loader({ request }: LoaderFunctionArgs) {
  const abortController = new AbortController();
  abortController.signal.addEventListener("abort", () =>
    abortController.abort(),
  );

  // simplified code

  return eventStream(
    AbortSignal.any([request.signal, abortController.signal]),
    (send: SendFunction) => {
      function notify(event: EventType) {
        try {
          send({ event, data: new Date().toISOString() });
        } catch (cause) {
          if (
            cause instanceof Error &&
            !cause.message.includes("Controller is already closed")
          ) {
            console.error("Failed to send SSE event", cause);
          }
        }
      }

      // simplified code
    },
  );
}
```

**Comment by JasonColeyNZ** [comment]
Are you sure why, because remix hasn't needed this in the past, and RR7 doesn't need it either, works very well in fact with the implementation that remix-live-loader uses. It's only when hono is added that the issues appear. if you remove hono server it all works as expected again?

**Comment by rphlmr** [comment]
Yep. I have this issue with Hono since Vite.
I don't know why, maybe a difference between ExpressJS and Hono.
Since it fixes my issues, I haven't dig deeper.
I should try without Remix utils, just in case.

**Comment by JasonColeyNZ** [comment]
If all we are doing is adding a try/catch around the send, then surely we will have a memory leak?
I have a console that shows the emitter listenerCount on the loaders and this just seems to be going up and up.
emitter.listenerCount()
HMR seems to be the main cause here, each time a save occurs in a stream modification, the old emitters are not being freed.

**Comment by JasonColeyNZ** [comment]
Interesting I can't actually get the emitter count over 3, or two windows, my only concern is with the potential with many users hitting pages at the same time?

**Comment by JasonColeyNZ** [comment]
Yeah with the sample I made, with three windows I now have an emitter count of 6 on one of the slugs, so its just getting higher and higher. I did find somewhere that nextjs were having the same error and their 'solution' was to patch react itself.

https://github.com/vercel/next.js/pull/57011/commits/f2009f34b956e1d91eda27d695d233e8d75b7ea9
https://x.com/Kingsley_codes/status/1808578766375113027

I did read also that nextjs  14 had similar issues.

also undici were looking into this, I tried Node 22 but this didn't help
https://github.com/nodejs/undici/issues/1564

**Comment by rphlmr** [comment]
So, I tried your example and I was able to reproduce it.

We can better handle the error and maintain a proper number of connections with:
```ts
import { eventStream } from "remix-utils/sse/server";
import { emitter } from "./emitter.server";

export function createEventStream(request: Request, eventName: string) {
  const abortController = new AbortController();
  abortController.signal.addEventListener("abort", () =>
    abortController.abort()
  );

  return eventStream(
    AbortSignal.any([request.signal, abortController.signal]),
    (send) => {
      const handle = () => {
        try {
          console.log("listenerCount: ", emitter.listenerCount(eventName));

          send({
            data: String(Date.now()),
          });
        } catch (error) {
          if (
            error instanceof Error &&
            !error.message.includes("Controller is already closed")
          ) {
            console.error("Failed to send SSE event", error);
          } else {
            // Remove dead listeners
            emitter.removeListener(eventName, handle);
          }
        }
      };

      emitter.addListener(eventName, handle);

      return () => {
        emitter.removeListener(eventName, handle);
      };
    }
  );
}

```
